modules/eventlog/src/main/java/org/apache/ignite/internal/eventlog/impl/ConfigurationBasedChannelRegistry.java

```diff
@@ -0,0 +1,115 @@
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.stream.Collectors;
+import org.apache.ignite.configuration.NamedListView;
+import org.apache.ignite.configuration.notifications.ConfigurationListener;
+import org.apache.ignite.configuration.notifications.ConfigurationNotificationEvent;
+import org.apache.ignite.internal.eventlog.api.EventChannel;
+import org.apache.ignite.internal.eventlog.config.schema.ChannelView;
+import org.apache.ignite.internal.eventlog.config.schema.EventLogConfiguration;
+
+class ConfigurationBasedChannelRegistry implements ChannelRegistry {
+    private final ReadWriteLock guard;
+
+    private final Map<String, EventChannel> cache;
+
+    private final Map<String, Set<EventChannel>> typeCache;
+
+    private final SinkRegistry sinkRegistry;
+
+    ConfigurationBasedChannelRegistry(EventLogConfiguration cfg, SinkRegistry sinkRegistry) {
+        this.guard = new ReentrantReadWriteLock();
+        this.cache = new HashMap<>();
+        this.typeCache = new HashMap<>();
+        this.sinkRegistry = sinkRegistry;
+
+        cfg.channels().listen(new CacheUpdater());
+    }
+
+    @Override
+    public EventChannel getByName(String name) {
+        guard.readLock().lock();
+        try {
+            return cache.get(name);
+        } finally {
+            guard.readLock().unlock();
+        }
+    }
+
+    @Override
+    public Set<EventChannel> findAllChannelsByEventType(String igniteEventType) {
+        guard.readLock().lock();
+        try {
+            Set<EventChannel> result = typeCache.get(igniteEventType);
+            return result == null ? Set.of() : new HashSet<>(result);
+        } finally {
+            guard.readLock().unlock();
+        }
+    }
+
+    private class CacheUpdater implements ConfigurationListener<NamedListView<ChannelView>> {
+        @Override
+        public CompletableFuture<?> onUpdate(ConfigurationNotificationEvent<NamedListView<ChannelView>> ctx) {
+            NamedListView<ChannelView> newListValue = ctx.newValue();
+
+            guard.writeLock().lock();
+
+            try {
+                cache.clear();
+                typeCache.clear();
+
+                newListValue.forEach(view -> {
+                    if (view.enabled()) {
+                        EventChannel channel = createChannel(view);
+                        cache.put(view.name(), channel);
+                        for (String eventType : view.events()) {
+                            typeCache.computeIfAbsent(
+                                    eventType.trim(),
+                                    t -> new HashSet<>()
+                            ).add(channel);
+                        }
+                    }
+                });
+
+                return completedFuture(null);
+            } finally {
+                guard.writeLock().unlock();
+            }
```

Comment:
It will be better to create a new collection outside the lock, then just update the references under the lock.